Azure.ResourceManager.Core/Resources/Location.cs

```diff
@@ -2,98 +2,458 @@
+
+        private Location()
         {
-            Name = GetDefaultName(location);
-            CanonicalName = GetCanonicalName(location);
-            DisplayName = GetDisplayName(location);
         }
 
-        internal Location()
+        private Location(string location)
         {
+            switch (DetectNameType(location))
+            {
+                case NameType.Name:
+                    Name = location;
+                    CanonicalName = location;
+                    DisplayName = location;
+                    break;
+                case NameType.CanonicalName:
+                    Name = GetDefaultNameFromCanonicalName(location);
+                    CanonicalName = location;
+                    DisplayName = GetDisplayNameFromCanonicalName(location);
+                    break;
+                case NameType.DisplayName:
+                    Name = GetDefaultNameFromDisplayName(location);
+                    CanonicalName = GetCanonicalNameFromDisplayName(location);
+                    DisplayName = location;
+                    break;
+            }
         }
 
+        private enum NameType
+        {
+            Name,
+            CanonicalName,
+            DisplayName,
+        }
+
+        /// <summary>
+        /// Gets default Location object: West US.
+        /// </summary>
         public static ref readonly Location Default => ref WestUS;
 
-        public string Name { get; internal set; }
+        /// <summary>
+        /// Gets a location name consisting of only lowercase characters without white spaces or any separation character between words, e.g. "westus".
+        /// </summary>
+        public string Name { get; private set; }
+
+        /// <summary>
+        /// Gets a location canonical name consisting of only lowercase chararters with a '-' between words, e.g. "west-us".
+        /// </summary>
+        public string CanonicalName { get; private set; }
+
+        /// <summary>
+        /// Gets a location display name consisting of titlecase words or alphanumeric characters separated by whitespaces, e.g. "West US"
+        /// </summary>
+        public string DisplayName { get; private set; }
+
+        /// <summary>
+        /// Creates a new location implicitly from a string.
+        /// </summary>
+        /// <param name="other">String to be assigned in the Name, CanonicalName or DisplayName form.</param>
+        public static implicit operator Location(string other)
+        {
+            if (other == null)
+                return null;
 
-        public string CanonicalName { get; internal set; }
+            var normalizedName = NormalizationUtility(other);
+            Location value;
+            if (PublicCloudLocations.TryGetValue(normalizedName, out value))
+            {
+                return value;
+            }
 
-        public string DisplayName { get; internal set; }
+            return new Location(other);
```

Comment:
Yes, for the preview we will let an exception occur in that case.
This will be address in a future work item to get the available locations for a subscription.